### PR TITLE
Add JSON reader/writer and binjs_convert_from_json command (fixes #194)

### LINKED ISSUE #194: add a command to convert possibly-invalid JSON to binjs file
We need a way to create an invalid-file testcase reliably and automatically, which is usable across format version bump, to test implementation against invalid input efficiently.

I think XML is sufficient to represent most possible format-independent invalid structure,
such like SyntaxError or something.

If such XML can be converted to other format without performing error check, that can be used as a testcase.

### PR DESCRIPTION
**Pull request by arai-a**
Rewrite of #238.

Now this uses json-rs.
the precision issue in float would be addressed by `FromJSON for f64`,
and WTF-8 handling is going to be handled in other PR, that just applies escape/unescape on the stringified JSON.

Then, because of lifetime issue (I cannot have reference to JSON tree from TokenReader), I've added yet another stage in reading JSON,
that is, first it converts the JSON tree to the list of tokens. the token corresponds to TokenReader function call.
and then the reader uses the tokens to generate the result of each function call.

Also, writer now uses intermediate JSON tree,
and also tweaked the format a bit, to simplify the filter (now field names and field values are inside single array).
format details are in mod.rs.

### CODE REVIEW

File: src/bin/convert_from_json.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        .join()
+        .expect("Error in dedicated thread");
+}
+
+fn main_aux() {
+    env_logger::init();
+
+    let matches = App::new("Converter from BinAST JSON format to other BinAST format")
+        .author("arai <arai@mozilla.com>")
+        .about("Convert a BinAST JSON file to other BinAST format file. The JSON file is exported by binjs_encode with advanced json option")
```
Comment (Yoric) [inline src/bin/convert_from_json.rs]: Maybe "To obtain a BinAST JSON file, you may use binjs_encode with command `advanced json`."

File: crates/binjs_io/src/binjs_json/mod.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+//! A format for generating invalid file, to test decoder implementation.
+//!
+//! The expected usage is the following:
+//!
+//!   1. Convert .js file to this JSON format (write module)
+//!   2. Modify the JSON file, like the following:
+//!       - Remove a field from an interface
+//!       - Add a field to an interface
+//!       - Replace a field value
+//!   3. Convert JSON file to other BinAST format (read modules)
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/mod.rs]: oh, it should be read module (read.rs)

File: crates/binjs_io/src/binjs_json/mod.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+//!   4. Feed the possibly-invalid BinAST to decoder
+//!
+//! The JSON format is in the following structure:
+//!
+//! tagged tuple
+//! ```ignore
+//! {
+//!   "@TYPE": "tagged tuple",
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/mod.rs]: yes, only `<...>` are variable things

File: crates/binjs_io/src/binjs_json/mod.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+//!   "@INTERFACE": "<interface name>",
+//!   "@FIELD": [
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/mod.rs]: it was typo.
the code uses `@FIELDS`

File: crates/binjs_io/src/binjs_json/mod.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+//!
+//!   1. Convert .js file to this JSON format (write module)
+//!   2. Modify the JSON file, like the following:
+//!       - Remove a field from an interface
+//!       - Add a field to an interface
+//!       - Replace a field value
+//!   3. Convert JSON file to other BinAST format (read modules)
+//!   4. Feed the possibly-invalid BinAST to decoder
+//!
+//! The JSON format is in the following structure:
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/mod.rs]: null for nullable string uses the following.

```
{
  "@TYPE": "string",
  "@VALUE": null
}
```

I'll add some more examples.

File: crates/binjs_io/src/binjs_json/read.rs [hunk 6]
```diff
@@ -0,0 +1,811 @@
...
+
+    /// The list of tokens, converted from JSON.
+    tokens: VecDeque<Token>,
+
+    /// Just to consume `R` type.
+    phantom: PhantomData<R>,
+}
+impl<R: Read> Decoder<R> {
+    /// Read JSON from `source`, parse it, and convert it to tokens.
+    pub fn new(mut source: R) -> Result<Self, TokenReaderError> {
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/read.rs]: multipart and entropy also uses `new`
so they should be addressed at the same time.
I'll go with `new` and file issue.

File: crates/binjs_io/src/binjs_json/read.rs [hunk 7]
```diff
@@ -0,0 +1,811 @@
...
+            None => {
+                Err(TokenReaderError::GenericError(
+                    format!("Expected '{}', but no more token", kind)
+                ))
+            }
+        }
+    }
+
+    /// Get the next token and update the context.
+    fn next_unchecked(&mut self) -> Result<Option<Token>, TokenReaderError> {
```
[the remark this answers is not in the document]
Comment (arai-a) [reply crates/binjs_io/src/binjs_json/read.rs]: it doesn't throw error if there's no more token.
I'll add comment.

File: crates/binjs_io/src/binjs_json/read.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yoric) [reply crates/binjs_io/src/binjs_json/read.rs]: Sounds good!

**Pull request merged**